# Data Prepocessing

In [1]:
import numpy as np
import pandas as pd
import csv
from sklearn import preprocessing
from sklearn.decomposition import PCA

## Bank Dataset

In [2]:
# Read data from file

with open('bank/bank-full.csv', 'r') as csvfile:
    spamreader = csv.reader(csvfile, delimiter=';')
    df = list(spamreader)
    df = np.asarray(df)
    df = pd.DataFrame(data=df[1:], columns=df[0])
    

In [3]:
# Inspecting the unknown values in the dataset

jobs = df['job']
unknown_jobs = 0
for job in jobs:
    if job == 'unknown':
        unknown_jobs = unknown_jobs + 1
        
educations = df['education']
unknown_education = 0
for education in educations:
    if education == 'unknown':
        unknown_education = unknown_education + 1

defaults = df['default']
unknown_default = 0
for default in defaults:
    if default == 'unknown':
        unknown_default = unknown_default + 1

contacts = df['contact']
unknown_contact = 0
for contact in contacts:
    if contact == 'unknown':
        unknown_contact = unknown_contact + 1
    
pdays = df['pdays']
unknown_pday = 0
for pday in pdays:
    if pday == '-1':      # Treating never contacted as unknown
        unknown_pday = unknown_pday + 1

pouts = df['poutcome']
unknown_pouts = 0
for pout in pouts:
    if pout == 'unknown':
        unknown_pouts = unknown_pouts + 1
    

print("Unknown values\n")
print("Unknown jobs :", unknown_jobs)
print("Unknown education :", unknown_education)
print("Unknown contacts :", unknown_contact)
print("Unknown pdays :", unknown_pday)
print("Unknown pouts :", unknown_pouts)



Unknown values

Unknown jobs : 288
Unknown education : 1857
Unknown contacts : 13020
Unknown pdays : 36954
Unknown pouts : 36959


### Observations from Data Set

- Total number of rows: 45211
- Total number of attributes: 16 + output attribute (y)
- Yes samples - 5289
- No samples - 39922
- The number of yes samples are much smaller than the number of no samples indicating that the data set is highly skewed.

- Unknown values :-
    - job - 288
    - education - 1857
    - contact - 13020
    - pdays - 36954
    - poutcome - 36959
- A very large number of rows have pdays = -1 and poutcome = unknown. These represent clients that have not been contacted in the past

In [4]:
df = df.drop('duration', axis=1)             # Removing duration attribute

# Separating columns as numerical and categorical
numeric = ['age', 'balance', 'campaign', 'pdays', 'previous']
categorical = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'day', 'month', 'poutcome']


In [5]:
# Substituting pdays = 999 for pdays = -1
pdays = df['pdays']
pdays = pdays.replace('-1', '999')
df['pdays'] = pdays


# Substituting pdays = 999 and poutcome = failure for pdays = -1 and poutcome = unknown
# prev_list = ['pdays', 'poutcome']
# prev_df = df[prev_list]
# prev_df = prev_df.replace({'pdays':'-1', 'poutcome':'unknown'}, {'pdays':'999', 'poutcome':'failure'})
# df[prev_list] = prev_df


In [6]:
# Taking care of categorical values

# factorize assigns unique numeric values to each category
#for c in categorical:
#    df[c] = pd.factorize(df[c])[0]
    
# Use dummy attributes for categorical values
df = pd.get_dummies(df, columns=categorical, prefix=categorical)
df['y'],un = pd.factorize(df['y'])

#df[numeric] = df[numeric].astype('float')
#df[categorical] = df[categorical].astype('category')

df['y'] = df['y'].astype(int) # Converting otuput variable to integers


In [7]:
# Normalizing numerical values

scaler = preprocessing.StandardScaler()
rescaled_df = scaler.fit_transform(df[numeric])
df[numeric] = pd.DataFrame(rescaled_df, columns=numeric)

/Users/arunjoseph/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype object were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Users/arunjoseph/anaconda3/lib/python3.7/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype object were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [8]:
#Saving as pickle file

df.to_pickle('data/bank_data_dummies.pkl') # using dummies and replacing pdays only

In [9]:
# Principal Component Analysis

y_values = df.loc[:, ['y']].values
df.drop(columns=['y'], axis=1, inplace=True)

pca = PCA()
#pca = PCA(n_components=10)
components = pca.fit_transform(df)
var_ratios = pca.explained_variance_ratio_
#rint(var_ratios)

# Calculating the number of attributes for 80% variance
s=0
count = 0
for ratio in var_ratios:
    s = s + ratio
    count = count + 1
    if s >= 0.8:
        break
print(count)

# Calculating the number of attributes for 95% variance
s=0
count1 = 0
for ratio in var_ratios:
    s = s + ratio
    count1 = count1+ 1
    if s >= 0.95:
        break
print(count1)

#df = pd.DataFrame(data=components)
df['y'] = y_values

15
44


### Observations from PCA

- There were 15 attributes (after removing duration) subjected to PCA
- 80% of the variance can be captured by 2 attributes after PCA
- 95% of the variance can be captured by 5 attributes after PCA

In [10]:
df


age   balance  campaign     pdays  previous  job_admin.  \
0      1.606965  0.256419 -0.569351  0.466408 -0.251940           0   
1      0.288529 -0.437895 -0.569351  0.466408 -0.251940           0   
2     -0.747384 -0.446762 -0.569351  0.466408 -0.251940           0   
3      0.571051  0.047205 -0.569351  0.466408 -0.251940           0   
4     -0.747384 -0.447091 -0.569351  0.466408 -0.251940           0   
5     -0.559037 -0.371551 -0.569351  0.466408 -0.251940           0   
6     -1.218254 -0.300608 -0.569351  0.466408 -0.251940           0   
7      0.100181 -0.446762 -0.569351  0.466408 -0.251940           0   
8      1.606965 -0.407679 -0.569351  0.466408 -0.251940           0   
9      0.194355 -0.252657 -0.569351  0.466408 -0.251940           0   
10     0.006007 -0.358742 -0.569351  0.466408 -0.251940           1   
11    -1.124080 -0.319329 -0.569351  0.466408 -0.251940           1   
12     1.136095 -0.445449 -0.569351  0.466408 -0.251940           0   
13     1.606965 -0.424100 -0.569351  0.466408 -0.251940           0   
14     1.512791 -0.394213 -0.569351  0.466408 -0.251940           0   
15     0.947747 -0.372207 -0.569351  0.466408 -0.251940           0   
16     0.382703 -0.443150 -0.569351  0.466408 -0.251940           1   
17     1.512791 -0.430341 -0.569351  0.466408 -0.251940           0   
18     1.795313 -0.427713 -0.569351  0.466408 -0.251940           0   
19    -0.747384 -0.447419 -0.569351  0.466408 -0.251940           0   
20    -1.218254 -0.209960 -0.569351  0.466408 -0.251940           0   
21     1.418617 -0.191568 -0.569351  0.466408 -0.251940           0   
22    -0.841558 -0.439865 -0.569351  0.466408 -0.251940           0   
23    -1.500776 -0.430998 -0.569351  0.466408 -0.251940           0   
24    -0.088167 -0.447419 -0.569351  0.466408 -0.251940           0   
25     0.288529 -0.569598 -0.569351  0.466408 -0.251940           1   
26    -0.182341 -0.363668 -0.569351  0.466408 -0.251940           0   
27     1.041921 -0.410306 -0.569351  0.466408 -0.251940           0   
28     0.476877 -0.528215 -0.246560  0.466408 -0.251940           0   
29    -0.464863 -0.360384 -0.569351  0.466408 -0.251940           0   
...         ...       ...       ...       ...       ...         ...   
45181  0.476877  1.811893 -0.246560 -2.438859  1.050473           0   
45182 -0.653211 -0.403737 -0.246560 -2.211318  1.918749           0   
45183  2.737053 -0.341006 -0.569351 -2.511408  2.787025           0   
45184  2.077835  0.043593 -0.569351 -2.755437  1.918749           0   
45185  1.795313  0.950405 -0.569351 -2.524599  1.484611           0   
45186  1.701139  0.045235 -0.569351 -2.485027  0.616335           0   
45187 -0.841558 -0.063806 -0.569351  0.466408 -0.251940           0   
45188 -1.124080 -0.216529 -0.569351 -1.014256  0.616335           0   
45189 -1.500776 -0.382061 -0.569351 -2.524599  1.918749           0   
45190 -0.841558 -0.402752 -0.569351 -2.208020  1.050473           0   
45191  3.207923  0.803922 -0.569351 -2.224509  0.182198           0   
45192 -1.124080 -0.196166 -0.569351  0.466408 -0.251940           0   
45193 -1.218254 -0.395198 -0.246560 -2.719163  1.484611           0   
45194  1.701139 -0.402095 -0.246560 -2.211318  1.918749           0   
45195  2.548705 -0.071032 -0.569351 -2.211318  2.352887           0   
45196 -1.500776 -0.329839 -0.569351  0.466408 -0.251940           0   
45197 -0.464863  0.048848 -0.569351  0.466408 -0.251940           0   
45198 -0.370689  0.021587 -0.246560  0.466408 -0.251940           0   
45199 -0.653211  0.037024  0.076230 -1.080210  4.957714           0   
45200 -0.276515 -0.264480  0.399020  0.466408 -0.251940           0   
45201  1.136095 -0.255941 -0.569351 -2.221211  1.484611           0   
45202 -0.653211 -0.264480 -0.569351  0.466408 -0.251940           1   
45203 -1.689124 -0.410306 -0.569351  0.466408 -0.251940           0   
45204  3.019575  0.488624 -0.569351 -2.696079  3.221163           0   
45205 -1.500776 -0.281559 -0.246560 

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 81 columns):
age                    45211 non-null float64
balance                45211 non-null float64
campaign               45211 non-null float64
pdays                  45211 non-null float64
previous               45211 non-null float64
job_admin.             45211 non-null uint8
job_blue-collar        45211 non-null uint8
job_entrepreneur       45211 non-null uint8
job_housemaid          45211 non-null uint8
job_management         45211 non-null uint8
job_retired            45211 non-null uint8
job_self-employed      45211 non-null uint8
job_services           45211 non-null uint8
job_student            45211 non-null uint8
job_technician         45211 non-null uint8
job_unemployed         45211 non-null uint8
job_unknown            45211 non-null uint8
marital_divorced       45211 non-null uint8
marital_married        45211 non-null uint8
marital_single         45211 non-null uint8
e

In [12]:
df.describe()

age       balance      campaign         pdays      previous  \
count  4.521100e+04  4.521100e+04  4.521100e+04  4.521100e+04  4.521100e+04   
mean   5.222844e-16  3.070663e-15  1.254699e-14 -9.859880e-14  1.355210e-13   
std    1.000011e+00  1.000011e+00  1.000011e+00  1.000011e+00  1.000011e+00   
min   -2.159994e+00 -3.081149e+00 -5.693506e-01 -2.824689e+00 -2.519404e-01   
25%   -7.473845e-01 -4.237719e-01 -5.693506e-01  4.664082e-01 -2.519404e-01   
50%   -1.823406e-01 -3.002800e-01 -2.465603e-01  4.664082e-01 -2.519404e-01   
75%    6.652252e-01  2.158743e-02  7.622994e-02  4.664082e-01 -2.519404e-01   
max    5.091402e+00  3.309478e+01  1.944365e+01  4.664082e-01  1.191360e+02   

         job_admin.  job_blue-collar  job_entrepreneur  job_housemaid  \
count  45211.000000     45211.000000      45211.000000   45211.000000   
mean       0.114375         0.215257          0.032890       0.027427   
std        0.318269         0.411005          0.178351       0.163326   
min        0.000000         0.000000          0.000000       0.000000   
25%        0.000000         0.000000          0.000000       0.000000   
50%        0.000000         0.000000          0.000000       0.000000   
75%        0.000000         0.000000          0.000000       0.000000   
max        1.000000         1.000000          1.000000       1.000000   

       job_management      ...          month_mar     month_may     month_nov  \
count    45211.000000      ...       45211.000000  45211.000000  45211.000000   
mean         0.209197      ...           0.010551      0.304483      0.087810   
std          0.406740      ...           0.102174      0.460193      0.283022   
min          0.000000      ...           0.000000      0.000000      0.000000   
25%          0.000000      ...           0.000000      0.000000      0.000000   
50%          0.000000      ...           0.000000      0.000000      0.000000   
75%          0.000000      ...           0.000000      1.000000      0.000000   
max          1.000000      ...           1.000000      1.000000      1.000000   

          month_oct     month_sep  poutcome_failure  poutcome_other  \
count  45211.000000  45211.000000      45211.000000    45211.000000   
mean       0.016323      0.012807          0.108403        0.040698   
std        0.126718      0.112441          0.310892        0.197592   
min        0.000000      0.000000          0.000000        0.000000   
25%        0.000000      0.000000          0.000000        0.000000   
50%        0.000000      0.000000          0.000000        0.000000   
75%        0.000000      0.000000          0.000000        0.000000   
max        1.000000      1.000000          1.000000        1.000000   

       poutcome_success  poutcome_unknown             y  
count      45211.000000      45211.000000  45211.000000  
mean           0.033421          0.817478      0.116985  
std            0.179735          0.386278      0.321406  
min            0.000000          0.000000      0.000000  
25%            0.000000          1.000000      0.000000  
50%            0.000000          1.000000      0.000000  
75%            0.000000          1.000000      0.000000  
max            1.000000          1.000000      1.000000  

[8 rows x 81 columns]

In [13]:
#Saving as pickle file

df.to_pickle('data/bank_data_dummies_pca.pkl')    # using dummies and replacing pdays only and performing PCA


### Bank Additional Dataset

In [14]:
# Reading data from file

with open('bank-additional/bank-additional-full.csv', 'r') as csvfile:
    spamreader = csv.reader(csvfile, delimiter=';')
    df = list(spamreader)
    df = np.asarray(df)
    df = pd.DataFrame(data=df[1:], columns=df[0])

In [15]:
# Inspecting the unknown values in the dataset

jobs = df['job']
unknown_jobs = 0
for job in jobs:
    if job == 'unknown':
        unknown_jobs = unknown_jobs + 1
        
maritals = df['marital']
unknown_maritals = 0
for marital in maritals:
    if marital == 'unknown':
        unknown_maritals = unknown_maritals + 1

educations = df['education']
unknown_education = 0
for education in educations:
    if education == 'unknown':
        unknown_education = unknown_education + 1

defaults = df['default']
unknown_default = 0
for default in defaults:
    if default == 'unknown':
        unknown_default = unknown_default + 1

housing = df['housing']
unknown_housing = 0
for house in housing:
    if house == 'unknown':
        unknown_housing = unknown_housing + 1
    
loans = df['loan']
unknown_loans = 0
for loan in loans:
    if loan == 'unknown':
        unknown_loans = unknown_loans + 1
    
pdays = df['pdays']
unknown_pday = 0
for pday in pdays:
    if pday == '999':      # Treating never contacted as unknown
        unknown_pday = unknown_pday + 1

pouts = df['poutcome']
unknown_pouts = 0
for pout in pouts:
    if pout == 'nonexistent':
        unknown_pouts = unknown_pouts + 1
    

print("Unknown values\n")
print("Unknown jobs :", unknown_jobs)
print("Unknown marital status :", unknown_maritals)
print("Unknown education :", unknown_education)
print("Unknown default :", unknown_default)
print("Unknown housing :", unknown_housing)
print("Unknown loans :", unknown_loans)
print("Unknown pdays :", unknown_pday)
print("Unknown pouts :", unknown_pouts)


Unknown values

Unknown jobs : 330
Unknown marital status : 80
Unknown education : 1731
Unknown default : 8597
Unknown housing : 990
Unknown loans : 990
Unknown pdays : 39673
Unknown pouts : 35563


### Observations from Data Set

- Total number of rows: 41188
- Total number of attributes: 19 + output attribute (y)
- Yes samples - 4640
- No samples - 36548

- The number of yes samples are much smaller than the number of no samples indicating that the data set is highly skewed.

- Unknown values :-
    - job - 288
    - education - 1857
    - contact - 13020
    - pdays - 36954
    - poutcome - 36959
- A very large number of rows have pdays = 999 and poutcome = nonexistent. These represent clients that have not been contacted in the past

In [16]:
df = df.drop('duration', axis=1)             # Removing duration attribute

numeric = ['age', 'campaign', 'pdays', 'previous', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed']
categorical = ['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome']


In [17]:
# Taking care of categorical values

# factorize assigns unique numeric values to each category
#for c in categorical:
#    df[c] = pd.factorize(df[c])[0]
    
# Use dummy attributes for categorical values
df = pd.get_dummies(df, columns=categorical, prefix=categorical)
df['y'],un = pd.factorize(df['y'])

#df[numeric] = df[numeric].astype('float')
#df[categorical] = df[categorical].astype('category')

#df['y'] = df['y'].astype(int) # Converting otuput variable to integers

In [18]:
# Normalizing numerical values

scaler = preprocessing.StandardScaler()
rescaled_df = scaler.fit_transform(df[numeric])
df[numeric] = pd.DataFrame(rescaled_df, columns=numeric)

/Users/arunjoseph/anaconda3/lib/python3.7/site-packages/sklearn/preprocessing/data.py:625: DataConversionWarning: Data with input dtype object were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/Users/arunjoseph/anaconda3/lib/python3.7/site-packages/sklearn/base.py:462: DataConversionWarning: Data with input dtype object were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [19]:
#Saving as pickle file

df.to_pickle('data/bank_additional_data_dummies.pkl') # using dummies and replacing pdays only
#df.to_pickle('data/bank_additional_data_factorize.pkl') # using dummies and replacing pdays only

In [20]:
# Principal Component Analysis

y_values = df.loc[:, ['y']].values
df.drop(columns=['y'], axis=1, inplace=True)

#pca = PCA()
pca = PCA(n_components=10)
components = pca.fit_transform(df)
var_ratios = pca.explained_variance_ratio_
#rint(var_ratios)

# Calculating the number of attributes for 80% variance
s=0
count = 0
for ratio in var_ratios:
    s = s + ratio
    count = count + 1
    if s >= 0.8:
        break
print(count)

# Calculating the number of attributes for 95% variance
s=0
count1 = 0
for ratio in var_ratios:
    s = s + ratio
    count1 = count1+ 1
    if s >= 0.95:
        break
print(count1)

df = pd.DataFrame(data=components)
df['y'] = y_values

10
10


### Observations from PCA

- There were 18 attributes (after removing duration) subjected to PCA
- 80% of the variance can be captured by 6 attributes after PCA
- 95% of the variance can be captured by 10 attributes after PCA


In [21]:
df

0         1         2         3         4         5         6  \
0     -1.410809  1.044812 -1.504958 -0.035326 -0.674517  0.941939  0.555259   
1     -1.492236  1.062986 -1.538064  0.062224 -0.801406  1.042715  0.565375   
2     -1.359959  0.434184 -0.130096 -0.921027 -0.061031  0.851891 -0.709292   
3     -1.392443  0.580186 -0.350896 -0.819978 -0.098846  0.968507  0.577221   
4     -1.401831  0.979199 -1.358399 -0.096713 -0.634324  0.944065  0.472528   
5     -1.497945  0.744934 -0.782935 -0.430400 -0.483526  1.123078  0.573568   
6     -1.397250  1.094451 -1.608777 -0.010725 -0.636896  0.830082  0.540212   
7     -1.503122  0.653106 -0.550797 -0.575067 -0.396202  1.173969  0.586207   
8     -1.315879  0.013641  0.969930 -1.719285  0.566959  0.784015 -0.745911   
9     -1.306606  0.004922  0.975820 -1.650383  0.483914  0.896159 -0.732150   
10    -1.503122  0.653106 -0.550797 -0.575067 -0.396202  1.173969  0.586207   
11    -1.306606  0.004922  0.975820 -1.650383  0.483914  0.896159 -0.732150   
12    -1.356348  0.138486  0.726778 -1.440681  0.293744  1.132057  0.475911   
13    -1.354282  0.985357 -1.467090 -0.049890 -0.630383  0.695501 -0.766829   
14    -1.379350  0.385815 -0.059959 -0.947852 -0.082032  0.916844 -0.707277   
15    -1.451449  0.975179 -1.468095  0.007580 -0.748224  0.823060 -0.843332   
16    -1.379350  0.385815 -0.059959 -0.947852 -0.082032  0.916844 -0.707277   
17    -1.472797  0.723359 -0.881262 -0.345631 -0.543715  0.920693 -0.814902   
18    -1.386254  0.765824 -1.024307 -0.286187 -0.546817  0.842019 -0.828048   
19    -1.441527  0.490490 -0.109248 -0.894802 -0.109564  1.102574  0.544691   
20    -1.389019  0.305923  0.325093 -1.232112  0.149088  1.086045  0.597806   
21    -1.477908  0.994854 -1.536942  0.078549 -0.839649  0.901450 -0.738227   
22    -1.302779  0.846743 -1.087245 -0.310079 -0.415872  0.719152 -0.784913   
23    -1.315226  0.444263 -0.093961 -0.987717  0.053850  0.740029 -0.758811   
24    -1.348249  0.439607 -0.116767 -0.967854  0.018479  0.779778 -0.716920   
25    -1.388373  0.450280 -0.032784 -1.086887  0.122536  0.854477  0.495241   
26    -1.373833  1.061948 -1.653429  0.023048 -0.676912  0.645980 -0.751953   
27    -1.493264  0.594866 -0.410605 -0.691659 -0.292360  1.132461  0.580326   
28    -1.469794  0.906782 -1.022331 -0.091257 -0.154709  0.847127  0.522469   
29    -1.462043  1.021904 -1.476891 -0.065258 -0.639559  0.801448 -0.138824   
...         ...       ...       ...       ...       ...       ...       ...   
41158  1.791684 -1.053882  1.291410  0.674590 -1.767799  1.690459  0.187555   
41159  5.572662  6.041274  4.942134  1.211049 -2.790208 -0.756954 -0.252978   
41160  1.806321 -1.093725  1.393046  0.530552 -1.604153  1.597581  0.203077   
41161  2.628832 -0.064727  2.035355  0.731899 -1.908974  0.961448 -0.419081   
41162  3.674598  3.927820  1.596219  2.259993 -2.746659  0.264551 -1.682023   
41163  4.288300  4.182044  3.972094  1.448039 -1.837633 -0.083340 -1.194310   
41164  3.661416  3.791798  2.070722  1.856652 -2.382896  0.334752 -0.380300   
41165  1.779141 -0.996635  1.217960  0.995790 -1.527314  1.608610  0.173553   
41166  2.366348  0.133956  2.182303  0.640793 -1.911361  2.014293  0.774158   
41167  1.772280 -1.171193  1.586509  1.015022 -1.093635  1.412210 -1.105267   
41168  1.759314 -0.963144  1.105813  1.059384 -1.594608  1.635539  0.190161   
41169  1.671186 -0.392850 -0.236052  2.936239 -1.680407  1.474377 -1.131168   
41170  4.436395  2.898494  3.376326  1.715664 -2.612254 -0.037386  0.966230   
41171  1.608486 -0.959303  1.422706  0.597681 -1.749948  2.335547 -1.226123   
41172  1.898869 -1.320854  1.848528  0.259976 -1.346880  1.350679 -1.159916   
41173  3.238129  1.663401  0.600187  2.238610 -3.099931  0.508952 -0.013871   
41174  6.816310  8.693942  4.170214  2.717326 -4.058269 -1.638306  0.629734   
41175  3.283657  0.786119  2.832376  0.825827 -2.141693  0.759050 -0.011435   
41176  1.846832 -1.079888  1.199165  0.810528 -1.839566  1.46681

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 11 columns):
0    41188 non-null float64
1    41188 non-null float64
2    41188 non-null float64
3    41188 non-null float64
4    41188 non-null float64
5    41188 non-null float64
6    41188 non-null float64
7    41188 non-null float64
8    41188 non-null float64
9    41188 non-null float64
y    41188 non-null int64
dtypes: float64(10), int64(1)
memory usage: 3.5 MB


In [23]:
df.describe()

0             1             2             3             4  \
count  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04   
mean  -4.900529e-16 -8.014102e-16  5.123501e-16 -7.474086e-16 -4.265145e-16   
std    2.028186e+00  1.196995e+00  1.080178e+00  1.002392e+00  9.596998e-01   
min   -3.364378e+00 -2.556008e+00 -6.489795e+00 -2.721863e+00 -4.058269e+00   
25%   -1.475767e+00 -6.181078e-01 -6.142255e-01 -6.234621e-01 -6.072588e-01   
50%   -1.293096e+00  4.761588e-02  4.361805e-02 -9.751166e-02 -1.314451e-01   
75%    2.087327e+00  4.749097e-01  6.633251e-01  4.719449e-01  4.484624e-01   
max    7.426491e+00  9.846704e+00  6.705562e+00  1.334323e+01  1.308642e+01   

                  5             6             7             8             9  \
count  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04  4.118800e+04   
mean  -6.116982e-16 -6.913961e-17  3.950796e-17  1.361929e-16 -1.968794e-17   
std    8.361476e-01  6.927227e-01  6.838179e-01  6.372338e-01  5.554882e-01   
min   -2.690196e+00 -1.858373e+00 -2.913930e+00 -2.628959e+00 -1.221840e+00   
25%   -8.305326e-01 -6.708533e-01 -3.552031e-01 -4.451111e-01 -4.821907e-01   
50%    5.805031e-02 -4.388652e-02 -5.675053e-02 -2.404736e-02  7.376846e-02   
75%    7.611406e-01  6.125531e-01  2.465669e-01  4.331854e-01  4.125179e-01   
max    2.893526e+00  2.837619e+00  7.226832e+00  2.831044e+00  1.833612e+00   

                  y  
count  41188.000000  
mean       0.112654  
std        0.316173  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [24]:
#Saving as pickle file

df.to_pickle('data/bank_additional_data_dummies_pca.pkl')    # using dummies and replacing pdays only and performing PCA